iothub_client/samples/pnp/pnp_temperature_controller/pnp_thermostat_component.c

```diff
@@ -0,0 +1,367 @@
+    {
+        LogError("Unable to build reported property response");
+    }
+    else
+    {
+        const char* jsonToSendStr = STRING_c_str(jsonToSend);
+        size_t jsonToSendStrLen = strlen(jsonToSendStr);
+
+        if ((iothubClientResult = IoTHubDeviceClient_SendReportedState(deviceClient, (const unsigned char*)jsonToSendStr, jsonToSendStrLen, NULL, NULL)) != IOTHUB_CLIENT_OK)
+        {
+            LogError("Unable to send reported state, error=%d", iothubClientResult);
+        }
+        else
+        {
+            LogInfo("Sending acknowledgement of property to IoTHub for component=%s", pnpThermostatComponent->componentName);
+        }
+    }
+
+    STRING_delete(jsonToSend);
+}
+
+void PnP_TempControlComponent_Report_MaxTempSinceLastReboot_Property(PNP_THERMOSTAT_COMPONENT_HANDLE pnpThermostatComponentHandle, IOTHUB_DEVICE_CLIENT_HANDLE deviceClient)
+{
+    PNP_THERMOSTAT_COMPONENT* pnpThermostatComponent = (PNP_THERMOSTAT_COMPONENT*)pnpThermostatComponentHandle;
+    char maximumTemperatureAsString[32];
+    IOTHUB_CLIENT_RESULT iothubClientResult;
+    STRING_HANDLE jsonToSend = NULL;
+
+    if (snprintf(maximumTemperatureAsString, sizeof(maximumTemperatureAsString), g_maxTempSinceLastRebootPropertyFormat, pnpThermostatComponent->maxTemperature) < 0)
+    {
+        LogError("Unable to create max temp since last reboot string for reporting result");
+    }
+    else if ((jsonToSend = PnPHelper_CreateReportedProperty(pnpThermostatComponent->componentName, g_maxTempSinceLastRebootPropertyName, maximumTemperatureAsString)) == NULL)
+    {
+        LogError("Unable to build max temp since last reboot property");
+    }
+    else
+    {
+        const char* jsonToSendStr = STRING_c_str(jsonToSend);
+        size_t jsonToSendStrLen = strlen(jsonToSendStr);
+
+        if ((iothubClientResult = IoTHubDeviceClient_SendReportedState(deviceClient, (const unsigned char*)jsonToSendStr, jsonToSendStrLen, NULL, NULL)) != IOTHUB_CLIENT_OK)
+        {
+            LogError("Unable to send reported state, error=%d", iothubClientResult);
+        }
+        else
+        {
+            LogInfo("Sending maximumTemperatureSinceLastReboot property to IoTHub for component=%s", pnpThermostatComponent->componentName);
+        }
+    }
+
+    STRING_delete(jsonToSend);
+}
+
+void PnP_ThermostatComponent_ProcessPropertyUpdate(PNP_THERMOSTAT_COMPONENT_HANDLE pnpThermostatComponentHandle, IOTHUB_DEVICE_CLIENT_HANDLE deviceClient, const char* propertyName, JSON_Value* propertyValue, int version)
+{
+    PNP_THERMOSTAT_COMPONENT* pnpThermostatComponent = (PNP_THERMOSTAT_COMPONENT*)pnpThermostatComponentHandle;
+
+    if (strcmp(propertyName, g_targetTemperaturePropertyName) != 0)
+    {
+        LogError("Property=%s was requested to be changed but is not part of the thermostat interface definition", propertyName);
+    }
+    else
+    {
+        double targetTemperature = json_value_get_number(propertyValue);
```

Comment:
`json_value_get_number` does not warn of incorrect inputs and always returns the default value of zero. 

Device code should be prepared to validate twin data as untrusted input.

Reply:
Good catch.  Other places in code do a `json_value_get_type` to make sure it's actually a number.  Adding that here too.  (`pnp_simple_thermostat.c` equivalent code of this has same bug which I'm fixing here too.)